src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.DeclarationInfo.cs

```diff
@@ -0,0 +1,307 @@
+                    || IsPropertyDeclaration(token, semanticModel, position, cancellationToken, out result)
+                    || IsPossibleVariableOrLocalMethodDeclaration(token, semanticModel, position, cancellationToken, out result))
+                {
+                    return result;
+                }
+
+                return default(NameDeclarationInfo);
+            }
+
+            private static bool IsPossibleVariableOrLocalMethodDeclaration(SyntaxToken token, SemanticModel semanticModel,
+                int position, CancellationToken cancellationToken, out NameDeclarationInfo result)
+            {
+                result = IsLastTokenOfType<ExpressionStatementSyntax>(
+                    token, semanticModel,
+                    e => e.Expression,
+                    unused => default(SyntaxTokenList),
+                    unused => ImmutableArray.Create(SymbolKind.Local),
+                    cancellationToken);
+                return result.Type != null;
+            }
+
+            private static bool IsPropertyDeclaration(SyntaxToken token, SemanticModel semanticModel, int position, CancellationToken cancellationToken, out NameDeclarationInfo result)
+            {
+                result = IsLastTokenOfType<PropertyDeclarationSyntax>(token, semanticModel, m => m.Type, m => m.Modifiers, GetPossibleDeclarations, cancellationToken);
+                return result.Type != null;
+            }
+
+            private static bool IsMethodDeclaration(SyntaxToken token, SemanticModel semanticModel, int position, CancellationToken cancellationToken, out NameDeclarationInfo result)
+            {
+                result = IsLastTokenOfType<MethodDeclarationSyntax>(token, semanticModel, m => m.ReturnType, m => m.Modifiers, GetPossibleDeclarations, cancellationToken);
+                return result.Type != null;
+            }
+
+            private static NameDeclarationInfo IsFollowingTypeOrComma<TSyntaxNode>(SyntaxToken token,
+                SemanticModel semanticModel,
+                Func<TSyntaxNode, SyntaxNode> typeSyntaxGetter,
+                Func<TSyntaxNode, SyntaxTokenList?> modifierGetter,
+                Func<DeclarationModifiers, ImmutableArray<SymbolKind>> possibleDeclarationComputer,
+                CancellationToken cancellationToken) where TSyntaxNode : SyntaxNode
+            {
+                if (!IsValidToken(token))
+                {
+                    return default(NameDeclarationInfo);
+                }
+
+                var target = token.GetAncestor<TSyntaxNode>();
+                if (target == null)
+                {
+                    return default(NameDeclarationInfo);
+                }
+
+                if (token.IsKind(SyntaxKind.CommaToken) && token.Parent != target)
+                {
+                    return default(NameDeclarationInfo);
+                }
+
+                var typeSyntax = typeSyntaxGetter(target);
```

Comment:
why are you not checking if TypeSyntax is null?  You check that in all the callers right?